### PR TITLE
minor error advisory spike

### PR DESCRIPTION
**Pull request by ripienaar**
far from complete but I am thinking of ways to notify people about
run time issues they would not know about because its either in
server logs or the detail sent to clients is (correctly) sparse

So we keep logging but also send them off as an advisory

```json
{
  "schema": "io.nats.jetstream.advisory.v1.error",
  "id": "KAdIvb73MxnoNDkrV4obzz",
  "timestamp": "2020-01-22T12:04:25.056937Z",
  "error": "message size exceeds maximum allowed",
  "stream": "test",
  "Server": "",
  "detail": {
    "account": "$G",
    "size": 8,
    "user": ""
  }
}
```

### CODE REVIEW

**Comment by ripienaar** [comment]
not for merging, just to start a discussion around approach

**Comment by derekcollison** [comment]
I like these. We should match client detail for all events from what we have now for api audit events.
For this one we should include a snippet of the message maybe too? But limit it in size to not blow things up IMO. Probably also want an error code as well as description for easier checking.

**Comment by ripienaar** [comment]
For what I did I use the error string for the snippet, we can always use that. 

Re error codes/description, I thought of that but we have none now and this is an annoying slipperly slope ime but I agree it would be nice to have so we should think how that might look

**Comment by ripienaar** [comment]
So I reworked this around one possible way to get error codes, descriptions and such in here, its ok I guess.

When I did this sort of thing in Ruby, here was the things I hate and I still hate it here too:

```go
seq, err = strconv.ParseUint(string(msg), 10, 64)
if err != nil {
   event := jsBadSequenceError.Wrap(err).Detail(JSErrorDetail{"account": c.acc.Name, "msg": msg})
   c.Debugf(event.LogLine())
   mset.sendq <- &jsPubMsg{reply, _EMPTY_, _EMPTY_, event.UserResponse(), nil, 0}
   return
}
```

When I skim the above, I dont read the code, I can see you got error handling and the error message tells me enough of what you were trying.  Here the message is lost. This is one of the most annoying things about idiomatic go, all these constants and returning constants and logging constants and sending constants, they dont stand out enough to me.

In constrast here, its clear - the message is still visible.

```go
metafile := path.Join(tdir, fi.Name(), JetStreamMetaFile)
metasum := path.Join(tdir, fi.Name(), JetStreamMetaFileSum)
buf, err := ioutil.ReadFile(metafile)
errEvent := jsStartUpError.Detail(JSErrorDetail{"metafile": metafile, "metasum": metasum})

if err != nil {
   a.sendJetStreamErrorAdvisory(s, errEvent.UserMessage("Error reading StreamTemplate metafile").Wrap(err))
   continue
}
```

So when I did this before in code basis I made the UserMessage() required in every use and so it was better.  Modern IDEs help, but still it annoyed me.

Anyway, let me know what you think about this.

```json
[#2] Received on [$JS.EVENT.ADVISORY.ERROR]: '{
  "schema": "io.nats.jetstream.advisory.v1.error",
  "id": "6dWdn1Od0EljiED2MJUOiZ",
  "timestamp": "2020-01-23T12:13:27.045032Z",
  "code": "JS5",
  "detail": {
    "checksum": "09763660ae4d147e",
    "metafile": "/var/folders/3v/s3_fnrkx54ndj070n7q4nw8c0000gn/T/jetstream/$G/streams/test/meta.inf",
    "metasum": "/var/folders/3v/s3_fnrkx54ndj070n7q4nw8c0000gn/T/jetstream/$G/streams/test/meta.sum",
    "sum": "6138b1f03e61fb88"
  },
  "description": "Errors encountered during JetStream startup for an account",
  "server": "NAXU4SSJGN7UMNSM672D3RL2M76SFW2IPSRGU3QZGVBQZFXDB2YCYLKH"
}'
```

File: server/jetstream.go [hunk 1]
```diff
@@ -331,34 +347,35 @@ func (a *Account) EnableJetStream(limits *JetStreamAccountLimits) error {
 			metafile := path.Join(tdir, fi.Name(), JetStreamMetaFile)
 			metasum := path.Join(tdir, fi.Name(), JetStreamMetaFileSum)
 			buf, err := ioutil.ReadFile(metafile)
+			errEvent := jsStartUpError.Detail(JSErrorDetail{"metafile": metafile, "metasum": metasum})
 			if err != nil {
-				s.Warnf("  Error reading StreamTemplate metafile %q: %v", metasum, err)
+				a.sendJetStreamErrorAdvisory(s, errEvent.UserMessage("Error reading StreamTemplate metafile").Wrap(err))
```
Comment (ripienaar) [inline server/jetstream.go]: these of course fire events super early like before clients are likely to be there.  But if you make a stream that catch them and it gets loaded early it will catch these so its quite good (assuming its not the one failing :))

We might consider sorting the order streams are loaded in so its deterministic and one can arrange for a audit stream to exist first.

File: server/jetstream_error.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	"time"
+
+	"github.com/nats-io/nuid"
+)
+
+var jsLimitsExceededError = NewJSError("JS1", "A message received exceeded Stream storage limits and were dropped").UserMessage("message size exceeds maximum allowed")
+var jsStorageWriteError = NewJSError("JS2", "A message could not be stored in the JetStream storage system").UserMessage("failed to store a msg")
+var jsBadSequenceError = NewJSError("JS3", "A invalid sequence number were requested").UserMessage("bad sequence argument")
+var jsStorageReadError = NewJSError("JS4", "Reading from the JetStream storage system failed").UserMessage("could not load message from storage")
+var jsStartUpError = NewJSError("JS5", "Errors encountered during JetStream startup for an account")
```
Comment (ripienaar) [inline server/jetstream_error.go]: its annoying that these like `jsStartupError` are defined here and not in the file they are needed - but if you do it other places the code ordering is easy to mess up.  Not really a fan of this code pattern.

Previous projects I made tooling and a place on the doc side where these codes can be looked up and they hold info about what they mean and how to fix, was ok, but too much work imo.

**Comment by philpennock** [comment]
Should there be a sampling rate for these error messages, or should they always be 1:1 with the bad data?

(nit: `and was dropped` not `and were dropped`)

**Comment by ripienaar** [comment]
These are audit trails for access to the admin interface, not a sample, so should be 1:1.

We have other kinds of messages that are samples, these are not.

**Comment by ripienaar** [comment]
D'oh sorry - also working on audit trails those are already in - sampling probably not they are kinds of things that happens once at startup.

**Comment by ColinSullivan1** [comment]
wdyt about using the [CloudEvents](https://github.com/cloudevents/spec/blob/v1.0/spec.md) format?

**Comment by ripienaar** [comment]
> wdyt about using the [CloudEvents](https://github.com/cloudevents/spec/blob/v1.0/spec.md) format?

I like CE, Derek wasnt sure about all the extra baggage by default so we'd add it as a option I bet.

I did basic schemas for some of the events we already incorporated here https://github.com/nats-io/jetstream/tree/master/schemas/jetstream  - obviously not final home.

**Comment by ColinSullivan1** [comment]
> > wdyt about using the [CloudEvents](https://github.com/cloudevents/spec/blob/v1.0/spec.md) format?
> 
> I like CE, Derek wasnt sure about all the extra baggage by default so we'd add it as a option I bet.
> 

+1 for an option or some way to translate in a service.  CE has become a popular format.

**Comment by ripienaar** [comment]
Yeah supporting CE is quite easy, just wrap these events in a CE header package and pull out a few key items, easy to do once we know exactly what we're doing

**Comment by ripienaar** [comment]
@philpennock I think rate limiting on some is a good idea, especially like write errors to the stores.  Often these will be sent from one stream and into another - we can quickly land in a infinite loop of one error triggerring another which in turn triggers another.

So specific errors can now be tagged as rate limited and we allow 10/second (or whatever we choose)

**Comment by ripienaar** [comment]
@derekcollison squashed

File: server/jetstream.go [hunk 3]
```diff
@@ -331,34 +351,35 @@ func (a *Account) EnableJetStream(limits *JetStreamAccountLimits) error {
...
 			buf, err := ioutil.ReadFile(metafile)
+			errEvent := jsStartUpError.Detail(JSErrorDetail{"metafile": metafile, "metasum": metasum})
 			if err != nil {
-				s.Warnf("  Error reading StreamTemplate metafile %q: %v", metasum, err)
+				a.sendJSError(s, errEvent.UserMessage("Error reading StreamTemplate metafile").Wrap(err))
 				continue
 			}
 			if _, err := os.Stat(metasum); os.IsNotExist(err) {
-				s.Warnf("  Missing StreamTemplate checksum for %q", metasum)
+				a.sendJSError(s, errEvent.UserMessage("Missing StreamTemplate checksum"))
```
Comment (kozlovic) [inline server/jetstream.go]: Is `.Wrap()` or `.Detail()` missing here or not?

File: server/stream.go [hunk 4]
```diff
@@ -666,3 +670,33 @@ func (mset *Stream) ackMsg(obs *Consumer, seq uint64) {
...
+	if event.ShouldRateLimit() && !errorEventLimit.Allow() {
+		return
+	}
+
+	e := event.Advisory()
+
+	mset.mu.Lock()
+	e.Stream = mset.config.Name
+	c.mu.Lock()
+	e.Server = mset.client.srv.Name()
```
Comment (kozlovic) [inline server/stream.go]: I assume that if you lock `c.mu.Lock()` it is because `c==mset.client`, right? But you can't call srv.Name() inside the client lock, that could create a lock inversion (Name() is locking *Server).
So you should write something like that instead:
```
e := event.Advisory()

mset.mu.RLock()
e.Stream = mset.config.Name
mset.mu.RUnlock()

c.mu.Lock()
srv := c.srv
c.mu.Unlock()
if srv != nil {
    e.Server = srv.Name()
}
```

.. and same comment than in Account.sendJSError() due to use of public API Server.Name() that requires server lock.

File: server/jetstream.go [hunk 5]
```diff
@@ -247,6 +248,25 @@ func (s *Server) getJetStream() *jetStream {
...
 
+func (a *Account) sendJSError(s *Server, event JSError) {
+	s.Warnf("  " + event.LogLine())
+
+	if event.ShouldRateLimit() && !errorEventLimit.Allow() {
+		return
+	}
+
+	advisory := event.Advisory()
+	advisory.Server = s.Name()
```
Comment (kozlovic) [inline server/jetstream.go]: Careful: s.Name() requires server lock. So make sure that where you call sendJSError, you are not under server lock or any lock that would cause lock inversion.
Alternatively, and since the server name is immutable, we could have a s.name() that just returns s.info.Name. If s.info is altered, we could if need be have a dedicated `name` field in the *Server structure and again have this immutable.

File: server/jetstream.go (hunk 3 above)
Comment (ripienaar) [reply server/jetstream.go]: No I dont think there's much value in wrapping the error from stat in this case, so just a basic error, we already know its missing and thats what the err is.

File: server/jetstream.go (hunk 5 above)
Comment (ripienaar) [reply server/jetstream.go]: Thanks will take a look - yeah the locking in server code is a bit hard to follow or to know what the model is, there seems no clear answer.  Many things - like name - that I think cant really change realistically at runtime is accessed under lock in other places.

If we decide to go with this pattern of code I'll have another look through these cases
